crypto/ml_dsa/ml_dsa_key.c

```diff
@@ -0,0 +1,380 @@
+/*
+ * Copyright 2024 The OpenSSL Project Authors. All Rights Reserved.
+ *
+ * Licensed under the Apache License 2.0 (the "License").  You may not use
+ * this file except in compliance with the License.  You can obtain a copy
+ * in the file LICENSE in the source distribution or at
+ * https://www.openssl.org/source/license.html
+ */
+
+#include <openssl/core_dispatch.h>
+#include <openssl/core_names.h>
+#include <openssl/params.h>
+#include <openssl/rand.h>
+#include "ml_dsa_local.h"
+#include "ml_dsa_key.h"
+#include "ml_dsa_params.h"
+#include "ml_dsa_matrix.h"
+
+/**
+ * @brief Create a new ML_DSA_KEY object
+ *
+ * @param libctx A OSSL_LIB_CTX object used for fetching algorithms.
+ * @param alg The algorithm name associated with the key type
+ * @returns The new ML_DSA_KEY object on success, or NULL on malloc failure
+ */
+ML_DSA_KEY *ossl_ml_dsa_key_new(OSSL_LIB_CTX *libctx, const char *alg)
+{
+    ML_DSA_KEY *ret;
+    const ML_DSA_PARAMS *params = ossl_ml_dsa_params_get(alg);
+
+    if (params == NULL)
+        return NULL;
+
+    ret = OPENSSL_zalloc(sizeof(*ret));
+    if (ret != NULL) {
+        if (!CRYPTO_NEW_REF(&ret->references, 1)) {
+            OPENSSL_free(ret);
+            return NULL;
+        }
+        ret->libctx = libctx;
+        ret->params = params;
+        vector_init(&ret->t0, params->k);
+        vector_init(&ret->t1, params->k);
+        vector_init(&ret->s2, params->k);
+        vector_init(&ret->s1, params->l);
+    }
+    return ret;
+}
+
+/**
+ * @brief Destroy a ML_DSA_KEY object
+ */
+void ossl_ml_dsa_key_free(ML_DSA_KEY *key)
+{
+    int i;
+
+    if (key == NULL)
+        return;
+
+    CRYPTO_DOWN_REF(&key->references, &i);
+    REF_PRINT_COUNT("ML_DSA_KEY", key);
+    if (i > 0)
+        return;
+    REF_ASSERT_ISNT(i < 0);
+
+    OPENSSL_free(key->pub_encoding);
+    OPENSSL_free(key->priv_encoding);
+    OPENSSL_free(key->propq);
+    CRYPTO_FREE_REF(&key->references);
+    OPENSSL_free(key);
+}
+
+/*
+ * @brief Increase the reference count for a ML_DSA_KEY object.
+ * @returns 1 on success or 0 otherwise.
+ */
+int ossl_ml_dsa_key_up_ref(ML_DSA_KEY *key)
+{
+    int i;
+
+    if (CRYPTO_UP_REF(&key->references, &i) <= 0)
+        return 0;
+
+    REF_PRINT_COUNT("ML_DSA_KEY", key);
+    REF_ASSERT_ISNT(i < 2);
+    return ((i > 1) ? 1 : 0);
+}
+
+/**
+ * @brief Are 2 keys equal?
+ *
+ * To be equal the keys must have the same key data and algorithm name.
+ *
+ * @param key1 A ML_DSA_KEY object
+ * @param key2 A ML_DSA_KEY object
+ * @param selection to select public and/or private component comparison.
+ * @returns 1 if the keys are equal otherwise it returns 0.
+ */
+int ossl_ml_dsa_key_equal(const ML_DSA_KEY *key1, const ML_DSA_KEY *key2,
+                          int selection)
+{
+    if (key1->params != key2->params)
+        return 0;
+    if (key1->pub_encoding != NULL) {
+        if (key2->pub_encoding == NULL
+                || memcmp(key1->pub_encoding, key1->pub_encoding,
```

Comment:
Let's fix the typo:
```suggestion
                || memcmp(key1->pub_encoding, key2->pub_encoding,
```
But also this does not work when one of the keys is just the private key, and the other just the public, but this is expected to work, e.g. for checking that public key in a key in a certificate matches a hypothetical corresponding private key.

The logic for comparing a public key against a private key seems to only be present in the *validate* method, which checks a single key with both parts present, but there is in fact a need to also be able to compare separate keys (as in the X.509 cert pub key against a separate private key (for which the public key was not, for some reason, explicitly loaded, unless that's expected to never happen in any relevant use-case).

So, unless I am mistaken, this should obtain any missing public key encoding from the private key on the fly.
In the ML-KEM case, the public key is never missing, it is always embedded in the private key. If that's too much additional space here, then perhaps temporary storage to compute the public key and discard it is the right thing to do.

Reply:
I am basing this on what ecx_match() does..

Reply:
I will have another look

Reply:
I will have to rethink this.. I would have to calculate key->tr (which is the encoded public key hash) when loading the public key and this depends on having a md_ctx which it doesnt have access to currently. I might decide to store the EVP_MD pointers in the key rather than a ctx..